### PR TITLE
refactor: Use BlockStatus to filter things

### PR DESCRIPTION
**Pull request by keroro520**
* refactor: Move `orphan_block_pool` into `SyncSharedState`
* feat: Rewrite `BlockStatus` and `block_status_map`
* perf: Use `BlockStatus` to avoid re-processing the same headers/blocks

### CODE REVIEW

File: sync/src/block_status.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
+use bitflags::bitflags;
+
+bitflags! {
+    pub struct BlockStatus: u32 {
+        const UNKNOWN                 =     0;
+
+        const HEADER_VERIFIED         =     1;
+        const BLOCK_RECEIVED          =     Self::HEADER_VERIFIED.bits | 2;
+        const BLOCK_STORED            =     Self::HEADER_VERIFIED.bits | Self::BLOCK_RECEIVED.bits | 16;
```
Comment (TheWaWaR) [inline sync/src/block_status.rs]: Should clear old stored block in case of fork.

File: sync/src/relayer/compact_block_process.rs [hunk 2]
```diff
@@ -43,6 +44,20 @@ impl<'a, CS: ChainStore + 'static> CompactBlockProcess<'a, CS> {
...
+        if status.contains(BlockStatus::BLOCK_STORED) {
+            return Ok(());
+        } else if status.contains(BlockStatus::BLOCK_INVALID) {
+            debug_target!(
+                crate::LOG_TARGET_RELAY,
+                "receive a compact block with invalid status, {:#x}, peer: {}",
+                block_hash,
+                self.peer
+            );
+            // TODO ban/punish this peer
```
Comment (TheWaWaR) [inline sync/src/relayer/compact_block_process.rs]: Punish should happen when insert `BLOCK_INVALID`.

**Comment by keroro520** [comment]
@nervos-bot-user try integration

File: sync/src/relayer/compact_block_process.rs (hunk 2 above)
Comment (keroro520) [reply sync/src/relayer/compact_block_process.rs]: We should punish the peer in 2 scenarios:

  * when inserts `BLOCK_INVALID`
  * receive a block already marked as `BLOCK_INVALID` before

But I think "ban" is dangerous, we should do it in another PR, but not here. Maybe after refactoring the scoring mechanism. So I just leave a TODO.

File: sync/src/block_status.rs (hunk 1 above)
Comment (keroro520) [reply sync/src/block_status.rs]: Discussed.

File: sync/src/block_status.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        vec![
+            BlockStatus::UNKNOWN,
+            BlockStatus::HEADER_VALID,
+            BlockStatus::BLOCK_RECEIVED,
+            BlockStatus::BLOCK_STORED,
+            BlockStatus::BLOCK_INVALID,
+        ]
+    }
+
+    fn assert_(includes: Vec<BlockStatus>, target: BlockStatus) {
```
Comment (doitian) [inline sync/src/block_status.rs]:
```suggestion
    fn assert_contain(includes: Vec<BlockStatus>, target: BlockStatus) {
```

File: sync/src/synchronizer/headers_process.rs [hunk 4]
```diff
@@ -393,49 +393,74 @@ where
 
     pub fn accept(&self) -> ValidationResult {
         let mut result = ValidationResult::default();
+
+        // FIXME If status == BLOCK_INVALID then return early. But which error
```
Comment (doitian) [inline sync/src/synchronizer/headers_process.rs]: Use a separate error code such as DuplicatedInvalidBlock or save the invalid reason as well.

Comment (keroro520) [reply sync/src/synchronizer/headers_process.rs]: I would like to persist a `recent_reject { hash => reason }` to record the invalid reasons.

**Pull request merged**